File: Data Structures & Algorithms/eating-bananas/submission-0.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left = 1
        right = max(piles)
        result = right

        while left <= right:
            mid = left + (right - left) // 2

            total_time = 0
            for pile in piles:
                total_time += math.ceil(pile / mid)
            
            if total_time <= h:
                result = mid
                right = mid - 1
            else:
                left = mid + 1
        
        return result
        

File: Data Structures & Algorithms/trapping-rain-water/submission-0.java
class Solution {
    public int trap(int[] height) {
        int left = 0;
        int right = height.length - 1;

        int maxLeft = height[left];
        int maxRight = height[right];

        int maxWater = 0;

        while(left < right) {
            if(maxLeft < maxRight) {
                left++;
                maxLeft = Math.max(maxLeft, height[left]);
                maxWater += maxLeft - height[left];
            } else {
                right--;
                maxRight = Math.max(maxRight, height[right]);
                maxWater += maxRight - height[right];
            }
        }
        return maxWater;
    }
}
